lib/diagnostics_channel.js

```diff
@@ -139,10 +212,165 @@ function hasSubscribers(name) {
+      this.error = error;
+    } else {
+      throw new ERR_INVALID_ARG_TYPE('nameOrChannels',
+                                     ['string', 'object', 'Channel'],
+                                     nameOrChannels);
+    }
+  }
+
+  subscribe(handlers) {
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      this[name]?.subscribe(handlers[name]);
+    }
+  }
+
+  unsubscribe(handlers) {
+    let done = true;
+
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      if (!this[name]?.unsubscribe(handlers[name])) {
+        done = false;
+      }
+    }
+
+    return done;
+  }
+
+  traceSync(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, error } = this;
+
+    try {
+      const result = start.runStores(ctx, fn, thisArg, ...args);
+      ctx.result = result;
+      return result;
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  tracePromise(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, asyncStart, asyncEnd, error } = this;
+
+    function reject(err) {
+      ctx.error = err;
+      error.publish(ctx);
+      asyncStart.publish(ctx);
+      // TODO: Is there a way to have asyncEnd _after_ the continuation?
+      asyncEnd.publish(ctx);
+      return PromiseReject(err);
+    }
+
+    function resolve(result) {
+      ctx.result = result;
+      asyncStart.publish(ctx);
+      // TODO: Is there a way to have asyncEnd _after_ the continuation?
+      asyncEnd.publish(ctx);
+      return result;
+    }
+
+    try {
+      const promise = start.runStores(ctx, fn, thisArg, ...args);
+      return PromisePrototypeThen(promise, resolve, reject);
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  traceCallback(fn, position = 0, ctx = {}, thisArg, ...args) {
+    const { start, end, asyncStart, asyncEnd, error } = this;
+
+    function wrappedCallback(err, res) {
+      if (err) {
+        ctx.error = err;
+        error.publish(ctx);
+      } else {
+        ctx.result = res;
+      }
+
+      asyncStart.publish(ctx);
+      try {
+        if (callback) {
+          return ReflectApply(callback, this, arguments);
+        }
+      } finally {
+        asyncEnd.publish(ctx);
+      }
+    }
+
+    const callback = ArrayPrototypeAt(args, position);
+    ArrayPrototypeSplice(args, position, 1, wrappedCallback);
+
+    try {
+      return start.runStores(ctx, fn, thisArg, ...args);
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+}
+
+function tracingChannel(nameOrChannels) {
+  return new TracingChannel(nameOrChannels);
+}
+
 module.exports = {
   channel,
   hasSubscribers,
   subscribe,
+  tracingChannel,
   unsubscribe,
-  Channel
+  Channel,
+  TracingChannel
```

Comment:
Theoretically one could sub-class it. Not sure why they'd want to though. 🤷🏻

Reply:
Similar as for `Channel` we export a factory method instead of telling users to use `new TracingChannel()`.
If we don't want construction via `new` allowing sub-classing sounds wrong.
Well one could use it for `instanceof`.
`ActiveChannel` is not exported and this is good to my understanding.
Also `Channel` should be not exported in my opinion but that ship has sailed.

Reply:
Channels will pass `instanceof` even when active due to the `SymbolHasInstance` method on `Channel`. Also, ideally I'd like to make the factory unnecessary. I have some ideas for sub-classing `TracingChannel` which I'd like to enable if at all possible.

Reply:
Maybe remove the factory `tracingChannel`? I see no reason to support both variants.

Reply:
Just went for that for consistency with Channel. We'll see how my sub-classing experiments go though. I'm not totally convinced of the utility of exposing both. I'll be taking another pass at cleaning stuff up soon, just busy with settling other work things before the holidays so might not get a chance to update this PR this week.

Reply:
That's fine. My main point is that it is always easier to add an export compared to removing it - even if it is experimental.

Reply:
Fair, I think I can just omit it then. It's easy to add later if I have a specific case for it.